File: frontend/src/api/accountLinkRequests.ts
/** AccountLinkRequest helpers (user-driven approval flow). */

import { apiClient } from './client';

export type AlrStatus = 'pending' | 'approved' | 'denied' | 'withdrawn';

export interface AccountLinkRequestRead {
  id: number;
  requester_user_id: number;
  physical_account_id: number;
  status: AlrStatus;
  request_note: string | null;
  decided_by_user_id: number | null;
  decided_at: string | null;
  decision_note: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateRequestPayload {
  physical_account_id: number;
  request_note?: string | null;
}

export interface DecisionPayload {
  decision_note?: string | null;
}

export async function createRequest(
  payload: CreateRequestPayload,
): Promise<AccountLinkRequestRead> {
  return (await apiClient.post<AccountLinkRequestRead>('/account-link-requests', payload)).data;
}

export async function listPending(): Promise<AccountLinkRequestRead[]> {
  return (await apiClient.get<AccountLinkRequestRead[]>('/account-link-requests')).data;
}

export async function listNeedsPush(): Promise<AccountLinkRequestRead[]> {
  return (await apiClient.get<AccountLinkRequestRead[]>('/account-link-requests/needs-push')).data;
}

export async function listMine(): Promise<AccountLinkRequestRead[]> {
  return (await apiClient.get<AccountLinkRequestRead[]>('/users/me/account-link-requests')).data;
}

export async function approve(
  id: number,
  payload: DecisionPayload = {},
): Promise<AccountLinkRequestRead> {
  return (
    await apiClient.post<AccountLinkRequestRead>(`/account-link-requests/${id}/approve`, payload)
  ).data;
}

export interface RetryPushResponse {
  request: AccountLinkRequestRead;
  key_push_outcome: Record<string, unknown>;
}

export async function retryPush(id: number): Promise<RetryPushResponse> {
  return (await apiClient.post<RetryPushResponse>(`/account-link-requests/${id}/retry-push`)).data;
}

export async function deny(
  id: number,
  payload: DecisionPayload = {},
): Promise<AccountLinkRequestRead> {
  return (
    await apiClient.post<AccountLinkRequestRead>(`/account-link-requests/${id}/deny`, payload)
  ).data;
}

export async function withdraw(id: number): Promise<AccountLinkRequestRead> {
  return (await apiClient.post<AccountLinkRequestRead>(`/account-link-requests/${id}/withdraw`))
    .data;
}

export interface RequesterRequestStats {
  total: number;
  approved: number;
  denied: number;
  withdrawn: number;
}

export interface KeyLateralSurface {
  fingerprint_sha256: string;
  label: string | null;
  physical_account_id: number;
  linux_username: string;
  server_hostname: string;
  pushed_at: string;
}

export interface RequestContext {
  request_id: number;
  requester_user_id: number;
  requester_username: string;
  requester_display_name: string;
  physical_account_id: number;
  linux_username: string;
  server_id: number;
  server_hostname: string;
  server_display_name: string | null;
  is_first_time_for_this_pa: boolean;
  requester_stats: RequesterRequestStats;
  requester_active_keys: Array<{
    id: number;
    fingerprint_sha256: string;
    key_type: string;
    comment: string;
  }>;
  lateral_surface: KeyLateralSurface[];
}

export async function fetchContext(id: number): Promise<RequestContext> {
  return (await apiClient.get<RequestContext>(`/account-link-requests/${id}/context`)).data;
}

File: backend/corelab_backend/api/v1/account_link_requests.py
"""``/api/v1/account-link-requests/*`` + ``/users/me/account-link-requests``."""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from ...auth_dependencies import (
    AuthenticatedUser,
    assert_server_admin,
    extract_request_context,
    get_current_user,
    list_granted_server_ids,
)
from ...db import get_session
from ...schemas.account_link import (
    AccountLinkRequestCreate,
    AccountLinkRequestDecision,
    AccountLinkRequestRead,
    AccountLinkRequestRetryPushResponse,
    RequestContext,
)
from ...services import account_link_request_service
from ...services.account_link_request_service import (
    AccountLinkRequestError,
    AlreadyLinkedError,
    DuplicatePendingError,
    InvalidStatusTransitionError,
    NoActiveSshKeyError,
    PhysicalAccountNotFoundError,
    RequestNotFoundError,
)

router = APIRouter(prefix="/account-link-requests", tags=["account-link-requests"])


async def _filter_rows_for_current_admin(
    session: AsyncSession,
    *,
    current: AuthenticatedUser,
    rows: list,
) -> list:
    if current.role == "lab_admin":
        return rows
    granted = set(await list_granted_server_ids(session, user=current))
    if not granted:
        return []
    from sqlalchemy import select as _sel

    from ...models import PhysicalAccount

    pa_ids = [r.physical_account_id for r in rows]
    if not pa_ids:
        return []
    pa_rows = (
        (await session.execute(_sel(PhysicalAccount).where(PhysicalAccount.id.in_(pa_ids))))
        .scalars()
        .all()
    )
    pa_server = {p.id: p.server_id for p in pa_rows}
    return [r for r in rows if pa_server.get(r.physical_account_id) in granted]


@router.post("", response_model=AccountLinkRequestRead, status_code=status.HTTP_201_CREATED)
async def create_request(
    payload: AccountLinkRequestCreate,
    request: Request,
    current: AuthenticatedUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> AccountLinkRequestRead:
    ip, ua = extract_request_context(request)
    try:
        row = await account_link_request_service.create_request(
            session,
            requester_user_id=current.id,
            physical_account_id=payload.physical_account_id,
            request_note=payload.request_note,
            lab_id=current.lab_id,
            request_ip=ip,
            user_agent=ua,
        )
    except PhysicalAccountNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except AlreadyLinkedError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    except DuplicatePendingError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    return AccountLinkRequestRead.model_validate(row)


@router.get("", response_model=list[AccountLinkRequestRead])
async def list_pending(
    current: AuthenticatedUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> list[AccountLinkRequestRead]:
    """Phase K — server admin sees pending across servers they manage;
    lab_admin sees all pending in the lab. Non-admins get an empty list
    (their own requests are reachable via /users/me/account-link-requests).
    """
    rows = await account_link_request_service.list_pending_in_lab(session, lab_id=current.lab_id)
    rows = await _filter_rows_for_current_admin(session, current=current, rows=list(rows))
    return [AccountLinkRequestRead.model_validate(r) for r in rows]


@router.get("/needs-push", response_model=list[AccountLinkRequestRead])
async def list_needs_push(
    current: AuthenticatedUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> list[AccountLinkRequestRead]:
    rows = await account_link_request_service.list_needing_key_push_in_lab(
        session, lab_id=current.lab_id
    )
    rows = await _filter_rows_for_current_admin(session, current=current, rows=list(rows))
    return [AccountLinkRequestRead.model_validate(r) for r in rows]


@router.get("/{request_id}/context", response_model=RequestContext)
async def request_context(
    request_id: int,
    current: AuthenticatedUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> RequestContext:
    """Signal bundle for the admin review card (Phase K K-5).
    Server admin can read context for requests on their granted servers."""
    try:
        ctx = await account_link_request_service.build_request_context(
            session, request_id=request_id, lab_id=current.lab_id
        )
    except RequestNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    await assert_server_admin(session, user=current, server_id=ctx["server_id"])
    return RequestContext.model_validate(ctx)


@router.post("/{request_id}/approve", response_model=AccountLinkRequestRead)
async def approve(
    request_id: int,
    payload: AccountLinkRequestDecision,
    request: Request,
    current: AuthenticatedUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> AccountLinkRequestRead:
    ip, ua = extract_request_context(request)
    # Look up server first so the server-admin gate can run.
    try:
        ctx = await account_link_request_service.build_request_context(
            session, request_id=request_id, lab_id=current.lab_id
        )
    except RequestNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    await assert_server_admin(session, user=current, server_id=ctx["server_id"])
    try:
        row = await account_link_request_service.approve_request(
            session,
            request_id=request_id,
            decision_note=payload.decision_note,
            admin_user_id=current.id,
            lab_id=current.lab_id,
            request_ip=ip,
            user_agent=ua,
        )
    except RequestNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except InvalidStatusTransitionError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    except NoActiveSshKeyError as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    return AccountLinkRequestRead.model_validate(row)


@router.post("/{request_id}/retry-push", response_model=AccountLinkRequestRetryPushResponse)
async def retry_push(
    request_id: int,
    request: Request,
    current: AuthenticatedUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> AccountLinkRequestRetryPushResponse:
    ip, ua = extract_request_context(request)
    try:
        ctx = await account_link_request_service.build_request_context(
            session, request_id=request_id, lab_id=current.lab_id
        )
    except RequestNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    await assert_server_admin(session, user=current, server_id=ctx["server_id"])
    try:
        row, outcome = await account_link_request_service.retry_push_for_request(
            session,
            request_id=request_id,
            admin_user_id=current.id,
            lab_id=current.lab_id,
            request_ip=ip,
            user_agent=ua,
        )
    except RequestNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except InvalidStatusTransitionError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    except NoActiveSshKeyError as exc:
        raise HTTPException(status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)) from exc
    return AccountLinkRequestRetryPushResponse(
        request=AccountLinkRequestRead.model_validate(row),
        key_push_outcome=outcome,
    )


@router.post("/{request_id}/deny", response_model=AccountLinkRequestRead)
async def deny(
    request_id: int,
    payload: AccountLinkRequestDecision,
    request: Request,
    current: AuthenticatedUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> AccountLinkRequestRead:
    ip, ua = extract_request_context(request)
    try:
        ctx = await account_link_request_service.build_request_context(
            session, request_id=request_id, lab_id=current.lab_id
        )
    except RequestNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    await assert_server_admin(session, user=current, server_id=ctx["server_id"])
    try:
        row = await account_link_request_service.deny_request(
            session,
            request_id=request_id,
            decision_note=payload.decision_note,
            admin_user_id=current.id,
            lab_id=current.lab_id,
            request_ip=ip,
            user_agent=ua,
        )
    except RequestNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except InvalidStatusTransitionError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    return AccountLinkRequestRead.model_validate(row)


@router.post("/{request_id}/withdraw", response_model=AccountLinkRequestRead)
async def withdraw(
    request_id: int,
    request: Request,
    current: AuthenticatedUser = Depends(get_current_user),
    session: AsyncSession = Depends(get_session),
) -> AccountLinkRequestRead:
    ip, ua = extract_request_context(request)
    try:
        row = await account_link_request_service.withdraw_request(
            session,
            request_id=request_id,
            requester_user_id=current.id,
            lab_id=current.lab_id,
            request_ip=ip,
            user_agent=ua,
        )
    except RequestNotFoundError as exc:
        raise HTTPException(status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    except InvalidStatusTransitionError as exc:
        raise HTTPException(status.HTTP_409_CONFLICT, detail=str(exc)) from exc
    except AccountLinkRequestError as exc:
        # not_requester (403) lands here.
        raise HTTPException(status.HTTP_403_FORBIDDEN, detail=str(exc)) from exc
    return AccountLinkRequestRead.model_validate(row)
